src/qibolab/_core/instruments/keysight/qcs.py

```diff
@@ -0,0 +1,172 @@
+"""Qibolab driver for Keysight QCS instrument set."""
+
+from collections import defaultdict
+from functools import reduce
+from typing import ClassVar, Optional
+
+import numpy as np
+from keysight import qcs
+
+from qibolab._core.components import AcquisitionChannel, Config, DcChannel, IqChannel
+from qibolab._core.execution_parameters import AveragingMode, ExecutionParameters
+from qibolab._core.identifier import ChannelId, Result
+from qibolab._core.instruments.abstract import Controller
+from qibolab._core.pulses import PulseId
+from qibolab._core.sequence import InputOps, PulseSequence
+from qibolab._core.sweeper import ParallelSweepers
+from qibolab._core.unrolling import Bounds
+
+from .pulse import (
+    process_acquisition_channel_pulses,
+    process_dc_channel_pulses,
+    process_iq_channel_pulses,
+)
+from .results import fetch_result, parse_result
+from .sweep import process_sweepers
+
+NS_TO_S = 1e-9
+
+BOUNDS = Bounds(
+    waveforms=1,
+    readout=1,
+    instructions=1,
+)
+
+__all__ = ["KeysightQCS"]
+
+
+class KeysightQCS(Controller):
+    """Driver for interacting with QCS controller server."""
+
+    bounds: str = "qcs/bounds"
+
+    # Map of QCS virtual channels to QCS physical channels
+    qcs_channel_map: qcs.ChannelMapper
```

Comment:
For dataclasses you can also document fields with docstrings, for example
```suggestion
    qcs_channel_map: qcs.ChannelMapper
    """Map of QCS virtual channels to QCS physical channels."""
```

Reply:
Changed in https://github.com/qiboteam/qibolab/pull/944/commits/402d266c4e2354dafb7d49c2c49888982763f2fd